File: superset-frontend/src/hooks/apiResources/lineage.ts
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { useApiV1Resource } from './apiResources';

// Database entity type
export type DatabaseEntity = {
  id: number;
  database_name: string;
  backend: string;
};

// Dataset entity type
export type DatasetEntity = {
  id: number;
  name: string;
  schema: string | null;
  table_name: string;
  database_id: number;
  database_name: string;
  chart_ids?: number[];
};

// Chart entity type
export type ChartEntity = {
  id: number;
  slice_name: string;
  viz_type: string;
  dashboard_ids?: number[];
  dataset_id?: number;
};

// Dashboard entity type
export type DashboardEntity = {
  id: number;
  title: string;
  slug: string;
  chart_ids?: number[];
};

// Dataset lineage response type
export type DatasetLineage = {
  dataset: DatasetEntity;
  upstream: {
    database: DatabaseEntity;
  };
  downstream: {
    charts: {
      count: number;
      result: ChartEntity[];
    };
    dashboards: {
      count: number;
      result: DashboardEntity[];
    };
  };
};

// Chart lineage response type
export type ChartLineage = {
  chart: ChartEntity & {
    datasource_id: number;
    datasource_type: string;
  };
  upstream: {
    dataset: DatasetEntity;
    database: DatabaseEntity;
  };
  downstream: {
    dashboards: {
      count: number;
      result: DashboardEntity[];
    };
  };
};

// Dashboard lineage response type
export type DashboardLineage = {
  dashboard: DashboardEntity & {
    published: boolean;
  };
  upstream: {
    charts: {
      count: number;
      result: ChartEntity[];
    };
    datasets: {
      count: number;
      result: DatasetEntity[];
    };
    databases: {
      count: number;
      result: DatabaseEntity[];
    };
  };
  downstream: null;
};

// A missing/empty identifier means we have nothing to fetch yet; skip the
// request so we never hit invalid endpoints like `/api/v1/chart//lineage`.
const isEmptyId = (idOrUuid: string | number): boolean =>
  idOrUuid === '' || idOrUuid == null;

/**
 * Hook to fetch lineage data for a dataset
 * @param idOrUuid Dataset ID or UUID
 * @param skip When true, defers the request (e.g. until the tab is active)
 */
export const useDatasetLineage = (idOrUuid: string | number, skip = false) =>
  useApiV1Resource<DatasetLineage>(
    `/api/v1/dataset/${idOrUuid}/lineage`,
    skip || isEmptyId(idOrUuid),
  );

/**
 * Hook to fetch lineage data for a chart
 * @param idOrUuid Chart ID or UUID
 * @param skip When true, defers the request (e.g. until the tab is active)
 */
export const useChartLineage = (idOrUuid: string | number, skip = false) =>
  useApiV1Resource<ChartLineage>(
    `/api/v1/chart/${idOrUuid}/lineage`,
    skip || isEmptyId(idOrUuid),
  );

/**
 * Hook to fetch lineage data for a dashboard
 * @param idOrSlug Dashboard ID or slug
 * @param skip When true, defers the request (e.g. until the tab is active)
 */
export const useDashboardLineage = (idOrSlug: string | number, skip = false) =>
  useApiV1Resource<DashboardLineage>(
    `/api/v1/dashboard/${idOrSlug}/lineage`,
    skip || isEmptyId(idOrSlug),
  );

File: tests/integration_tests/fixtures/lineage.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
import pytest

from superset import db
from superset.models.dashboard import Dashboard
from superset.models.slice import Slice
from superset.utils.database import get_example_database
from tests.integration_tests.dashboard_utils import create_table_metadata


@pytest.fixture
def lineage_test_data(app_context, load_birth_names_data):
    """
    Base fixture that creates a simple lineage structure and returns
    the created entities (database, dataset, charts, dashboard).
    """
    database = get_example_database()

    # Create dataset
    dataset = create_table_metadata(
        table_name="lineage_test_dataset",
        database=database,
    )
    db.session.add(dataset)
    db.session.flush()

    # Create charts
    chart1 = Slice(
        slice_name="Lineage Test Chart 1",
        viz_type="table",
        datasource_id=dataset.id,
        datasource_type="table",
        params="{}",
    )
    chart2 = Slice(
        slice_name="Lineage Test Chart 2",
        viz_type="pie",
        datasource_id=dataset.id,
        datasource_type="table",
        params="{}",
    )
    db.session.add(chart1)
    db.session.add(chart2)
    db.session.flush()

    # Create dashboard with charts
    dashboard = Dashboard(
        dashboard_title="Lineage Test Dashboard",
        slug="lineage-test-dashboard",
        slices=[chart1, chart2],
        published=True,
    )
    db.session.add(dashboard)
    db.session.commit()

    # Return the created entities
    result = {
        "database": database,
        "dataset": dataset,
        "charts": [chart1, chart2],
        "dashboard": dashboard,
    }

    yield result

    # Cleanup
    db.session.delete(dashboard)
    db.session.delete(chart1)
    db.session.delete(chart2)
    for col in dataset.columns + dataset.metrics:
        db.session.delete(col)
    db.session.delete(dataset)
    db.session.commit()


@pytest.fixture(autouse=False)
def inject_expected_dataset_lineage(request, lineage_test_data):
    """
    Injects dataset lineage data into test class instance.
    """
    dataset = lineage_test_data["dataset"]
    database = lineage_test_data["database"]
    charts = lineage_test_data["charts"]
    dashboard = lineage_test_data["dashboard"]

    request.instance.dataset_lineage = {
        "dataset_id": dataset.id,
        "expected": {
            "dataset": {
                "id": dataset.id,
                "name": dataset.name,
                "schema": dataset.schema,
                "table_name": dataset.table_name,
                "database_id": database.id,
                "database_name": database.database_name,
            },
            "upstream": {
                "database": {
                    "id": database.id,
                    "database_name": database.database_name,
                    "backend": database.backend,
                }
            },
            "downstream": {
                "charts": {
                    "count": 2,
                    "result": [
                        {
                            "id": charts[0].id,
                            "slice_name": charts[0].slice_name,
                            "viz_type": charts[0].viz_type,
                            "dashboard_ids": [dashboard.id],
                        },
                        {
                            "id": charts[1].id,
                            "slice_name": charts[1].slice_name,
                            "viz_type": charts[1].viz_type,
                            "dashboard_ids": [dashboard.id],
                        },
                    ],
                },
                "dashboards": {
                    "count": 1,
                    "result": [
                        {
                            "id": dashboard.id,
                            "title": dashboard.dashboard_title,
                            "slug": dashboard.slug,
                            "chart_ids": sorted([charts[0].id, charts[1].id]),
                        }
                    ],
                },
            },
        },
    }


@pytest.fixture(autouse=False)
def inject_expected_chart_lineage(request, lineage_test_data):
    """
    Injects chart lineage data into test class instance.
    """
    dataset = lineage_test_data["dataset"]
    database = lineage_test_data["database"]
    chart = lineage_test_data["charts"][0]  # Use first chart
    dashboard = lineage_test_data["dashboard"]

    request.instance.chart_lineage = {
        "chart_id": chart.id,
        "expected": {
            "chart": {
                "id": chart.id,
                "slice_name": chart.slice_name,
                "viz_type": chart.viz_type,
            },
            "upstream": {
                "dataset": {
                    "id": dataset.id,
                    "name": dataset.name,
                    "schema": dataset.schema,
                    "table_name": dataset.table_name,
                    "database_id": database.id,
                    "database_name": database.database_name,
                },
                "database": {
                    "id": database.id,
                    "database_name": database.database_name,
                    "backend": database.backend,
                },
            },
            "downstream": {
                "dashboards": {
                    "count": 1,
                    "result": [
                        {
                            "id": dashboard.id,
                            "title": dashboard.dashboard_title,
                            "slug": dashboard.slug,
                        }
                    ],
                }
            },
        },
    }


@pytest.fixture(autouse=False)
def inject_expected_dashboard_lineage(request, lineage_test_data):
    """
    Injects dashboard lineage data into test class instance.
    """
    dataset = lineage_test_data["dataset"]
    database = lineage_test_data["database"]
    charts = lineage_test_data["charts"]
    dashboard = lineage_test_data["dashboard"]

    request.instance.dashboard_lineage = {
        "dashboard_id": dashboard.id,
        "expected": {
            "dashboard": {
                "id": dashboard.id,
                "title": dashboard.dashboard_title,
                "slug": dashboard.slug,
                "published": dashboard.published,
            },
            "upstream": {
                "charts": {
                    "count": 2,
                    "result": [
                        {
                            "id": charts[0].id,
                            "slice_name": charts[0].slice_name,
                            "viz_type": charts[0].viz_type,
                            "dataset_id": dataset.id,
                        },
                        {
                            "id": charts[1].id,
                            "slice_name": charts[1].slice_name,
                            "viz_type": charts[1].viz_type,
                            "dataset_id": dataset.id,
                        },
                    ],
                },
                "datasets": {
                    "count": 1,
                    "result": [
                        {
                            "id": dataset.id,
                            "name": dataset.name,
                            "schema": dataset.schema,
                            "table_name": dataset.table_name,
                            "database_id": database.id,
                            "database_name": database.database_name,
                            "chart_ids": sorted([charts[0].id, charts[1].id]),
                        }
                    ],
                },
                "databases": {
                    "count": 1,
                    "result": [
                        {
                            "id": database.id,
                            "database_name": database.database_name,
                            "backend": database.backend,
                        }
                    ],
                },
            },
            "downstream": None,
        },
    }
